app/controllers/permissions_controller.ts

```diff
@@ -0,0 +1,226 @@
+    }
+    const userId = parsedUserId;
+
+    const targetUser = await User.find(userId);
+    if (targetUser === null) {
+      throw new NotFoundException("User not found");
+    }
+
+    const Model = resourceRegistry[resourceName];
+
+    if (resourceType === "class") {
+      await Acl.model(targetUser).allow(action, Model);
+      return { success: true };
+    }
+
+    const rawId = resource.id;
+    const parsedId =
+      typeof rawId === "number"
+        ? rawId
+        : typeof rawId === "string"
+          ? Number(rawId)
+          : Number.NaN;
+    if (!Number.isFinite(parsedId)) {
+      return { success: false, error: "Missing resource.id" };
+    }
+    const instance = await Model.find(parsedId);
+    if (instance === null) {
+      throw new NotFoundException("Resource not found");
+    }
+    await Acl.model(targetUser).allow(action, instance);
+    return { success: true };
+  }
+
+  async revoke({ request, auth }: HttpContext) {
+    await this.ensureSolvroAdmin(auth);
+
+    const body = request.only(["userId", "action", "resource"]) as Record<
+      string,
+      unknown
+    >;
+    const rawUserId = body.userId;
+    const parsedUserId =
+      typeof rawUserId === "number"
+        ? rawUserId
+        : typeof rawUserId === "string"
+          ? Number(rawUserId)
+          : Number.NaN;
+    const rawAction = body.action;
+    const rawResource = body.resource;
+
+    if (
+      !Number.isFinite(parsedUserId) ||
+      typeof rawAction !== "string" ||
+      typeof rawResource !== "object" ||
+      rawResource === null
+    ) {
+      return { success: false, error: "Invalid request body" };
+    }
+    if (
+      rawAction !== "read" &&
+      rawAction !== "create" &&
+      rawAction !== "update" &&
+      rawAction !== "destroy"
+    ) {
+      return { success: false, error: "Invalid action" };
+    }
+    const action = rawAction;
+    const resource = rawResource as {
+      type?: unknown;
+      name?: unknown;
+      id?: unknown;
+    };
+    const resourceType = resource.type;
+    const resourceName = resource.name;
+    if (resourceType !== "class" && resourceType !== "model") {
+      return { success: false, error: "Invalid resource.type" };
+    }
+    if (!isResourceKey(resourceName)) {
+      return { success: false, error: "Invalid resource.name" };
+    }
+    const userId = parsedUserId;
+
+    const targetUser = await User.find(userId);
+    if (targetUser === null) {
+      throw new NotFoundException("User not found");
+    }
+
+    const Model = resourceRegistry[resourceName];
+
+    // Try common revoke method names; if library exposes a different API, TS will guide fixes.
+    const manager = Acl.model(targetUser) as unknown as {
+      disallow?: (action: ActionSlug, target: unknown) => Promise<void>;
+      revoke?: (action: ActionSlug, target: unknown) => Promise<void>;
+      remove?: (action: ActionSlug, target: unknown) => Promise<void>;
+    };
```

Comment:
why do we even need these casts? isn't there a type available for this interface?